### DISCUSSION TITLE
[shared_ptr] Compile error with include "nanobind/stl/vector.h".

### DISCUSSION THREAD

**Discussion by huangweiwu** [wjakob/nanobind#269] [Q&A]
I'm binding a function witch return a shared_ptr with the code below.
It's ok when not include "nanobind/stl/vector.h".
But when I include "nanobind/stl/vector.h", it comes up with compile error.
Binding 'shared_ptr<T>' requires that 'T' can also be bound by nanobind. It appears that you specified a type which would undergo conversion/copying, which is not allowed.	
```
std::shared_ptr<std::vector<float>> test_func(){
    std::shared_ptr<std::vector<float>> a;
    return a;
}

NB_MODULE(my_ext, m) {
    m.def("test_func", &test_func);
}

**Comment by wjakob** [comment]
This is correct. You can handle `std::vector<>` either as an opaque class binding (with `stl/bind_vector.h`, avoids copies) or as a type caster (with `stl/vector.h`, always copies). The `std::shared_ptr` support is only compatible with class bindings since the desire is obviously not to copy. It cannot be mixed with type casters.

**Reply by wjakob** [reply]
Please review this page to understand the differences: https://nanobind.readthedocs.io/en/latest/exchanging.html

**Reply by huangweiwu** [reply]
Get it!
Furthermore, stl/bind_vector.h and stl/vector.h can not be included in the same source file, is that right?

**Reply by wjakob** [reply]
Normally one or the other. Or you can include both and use `NB_MAKE_OPAQUE` to selectively disable the type caster for a particular flavor of `std::vector`.

**Reply by huangweiwu** [reply]
Get it! Thanks!
